Docs for Google Cloud Platform (#9780)
solo-io/gloo: docs/content/guides/traffic_management/destination_types/google-cloud-run.md

---
title: Google Cloud Platform
weight: 100
description: Routing to Google Cloud Run as an Upstream.
---

Route traffic requests directly to a Google Cloud Platform (GCP) service, such as [Google Cloud Run](https://cloud.google.com/run).

{{% notice note %}}
{{< readfile file="static/content/enterprise_only_feature_disclaimer" markdown="true">}}
{{% /notice %}}

## Before you begin

> Make sure that your Google Cloud account is ready to use with the Cloud Run and Kubernetes Engine (GKE) services.

Reviewer: ```suggestion
Prepare your Google Cloud account to use it with the Cloud Run and Kubernetes Engine (GKE) services.
```

1. Identify a Google Cloud project with billing enabled that you plan to use for both the Cloud Run and Kubernetes Engine instances.
2. Verify that you have the appropriate permissions to create Cloud Run and Kubernetes Engine instances.

## Step 1: Set up your Google Cloud environment {#google-cloud}

Create a sample Cloud Run workload. Then, use Google Workload Identity to authorize Gloo Gateway to send requests to the Cloud Run workload.

1. In your Google Cloud project, identify or create a Kubernetes Engine cluster. For an example, follow the [Kubernetes Engine quickstart in the Google Cloud docs](https://cloud.google.com/kubernetes-engine/docs/quickstarts/create-cluster).

> 2. [Install Gloo Gateway Enterprise **version 1.17 or later** in your GKE cluster]({{% versioned_link_path fromRoot="/installation/platform_configuration/cluster_setup/#google-kubernetes-engine-gke" %}}).

Reviewer: ```suggestion
2. [Install Gloo Gateway Enterprise **version 1.17 or later** in your Kubernetes Engine cluster]({{% versioned_link_path fromRoot="/installation/platform_configuration/cluster_setup/#google-kubernetes-engine-gke" %}}).
```

3. In the same Google Cloud project as your cluster, deploy a `hello-world` Cloud Run application by following the [Deploy to Cloud Run quickstart in the Google Cloud docs](https://cloud.google.com/run/docs/quickstarts/deploy-container).

4. In your cluster, link the `gateway-proxy` Kubernetes ServiceAccount in the `gloo-system` namespace to your Google IAM service account. This way, the `gateway-proxy` can authenticate to your Google Cloud APIs by using Workload Identity Federation for GKE. 
   * At a minimum, the IAM service account must include the `run.invoker` and `iam.serviceAccountUser` roles.
   * For steps, see the [Kubernetes ServiceAccounts to IAM guide in the Google Cloud docs](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity#kubernetes-sa-to-iam).

## Step 2: Create the Upstream for the Google Cloud Run service {#upstream}

> The following examples create a basic VirtualService that routes to an Upstream for your Cloud Run workload.

Reviewer: ```suggestion
The following examples create a basic VirtualService that routes traffic to an Upstream that represents your Cloud Run workload.
```

> 1. Create an Upstream that represents your Cloud Run workload. Replace the `host` with the Cloud Run endpoint that your GKE cluster can access. For more options, see the [API docs]({{% versioned_link_path fromRoot="/reference/api/github.com/solo-io/gloo/projects/gloo/api/v1/enterprise/options/gcp/gcp.proto.sk/#upstreamspec" %}}).

Reviewer: not sure if there is maybe a doc that we can link here for how to get that URL?

Author: So if they follow the tutorial to create the Cloud Run, it is in that, but I can also add where it is in the GCP console.
